### PR TITLE
Jonavin merc additional encoder functionality

### PR DESCRIPTION
**Pull request by Jonavin**
## Description

  - FN layer has encoder dynamic selectable key codes and display on OLED
        
        Change these in keymap.c to assign your desired key selection

static const keycodedescType PROGMEM keyselection[] = { 
   // list of key codes that will be scrollled through by encoder and description    
        {"TASK",    KC_TASK},
        {"INS",     KC_INS},       
        {"DEL",     KC_DEL},
        {"PrtSc",   KC_PSCR},
        {"ScrLk",   KC_SCLN},
        {"Break",   KC_PAUS},
        {"C-A-D",   KC_CAD},  // Ctrl-Alt-Del 
        {"AltF4",   KC_AF4},
        {"PLAY",    KC_MEDIA_PLAY_PAUSE}
};

  - Additional encoder functionality 
        While holding Left Ctrl, navigates next or previous word
        While holding Left Alt, media next track or previous track

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/mechwild/mercutio/keymaps/jonavin/keymap.c [hunk 1]
```diff
@@ -56,30 +64,90 @@ const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
...
             oled_clear();
             oled_render();
         #endif
         switch (index) {
             case 0:         // This is the only encoder right now, keeping for consistency
-                if ( clockwise ) {
-                    if ( selected_layer  < 3 && keyboard_report->mods & MOD_BIT(KC_LSFT) ) { // If you are holding L shift, encoder changes layers
-                        selected_layer ++;
-                        layer_move(selected_layer);
+                switch(biton32(layer_state)){  // special handling per layer
```
Comment (drashna) [inline keyboards/mechwild/mercutio/keymaps/jonavin/keymap.c:104]:
```suggestion
                switch(get_highest_layer(layer_state)){  // special handling per layer
```

File: keyboards/mechwild/mercutio/keymaps/jonavin/keymap.c [hunk 2]
```diff
@@ -56,30 +64,90 @@ const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
   	KC_TRNS, KC_TRNS, KC_TRNS,          KC_BSPC, KC_TRNS,          KC_TRNS,          KC_TRNS, KC_TRNS,          KC_TRNS )
 };
 
+/*  These are needed whether encoder function is enabled or not when ENCFUNC keycode is pressed.
+    Defaults never changes if no encoder present to change it
+*/
+const uint16_t fkeycodes[] = { // list of key codes that will be scrollled through by encoder
+    KC_TASK, KC_INS, KC_DEL, KC_PSCR, KC_SCLN, KC_PAUS, KC_CAD, KC_AF4, KC_MEDIA_PLAY_PAUSE
+};
+const char* fkeydesc[] =    { // list of desc to be shown on LCD max 5 chars will be shown
```
Comment (drashna) [inline keyboards/mechwild/mercutio/keymaps/jonavin/keymap.c:73]: Best to use progmem, whenever possible for OLED rendering.  I think you need: 

```suggestion
static const PROGMEM char* fkeydesc[] =    { // list of desc to be shown on LCD max 5 chars will be shown
```

Comment (Jonavin) [reply keyboards/mechwild/mercutio/keymaps/jonavin/keymap.c]: The string array is not used to directly render OLED but used in a oled_write() call and dynamically changes. The whole array would have to be loaded into ram anyways. Would this not just complicate the need to read from PROGMEM and had no operational ram savings?

**Comment by Jonavin** [comment]
I rewrote the whole storage piece to use a PROGMEM typedef struct to reduce the memory footprint

**Review by drashna — APPROVED** [review]

File: keyboards/mechwild/mercutio/keymaps/jonavin/keymap.c [hunk 3]
```diff
@@ -147,14 +247,15 @@ const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
...
-                oled_write_P(PSTR("            "), false);
+                oled_write_P(PSTR("             "), false);
             } else {
                 switch (get_highest_layer(layer_state)) {
                     case 0:
                         oled_write_P(PSTR("Temp BASE"), false);
                         break;
                     case 1:
-                        oled_write_P(PSTR("Temp FN"), false);
+                        sprintf(fn_str, "Temp FN %5s", selectedkey_rec.keydesc);
```
Comment (drashna) [inline keyboards/mechwild/mercutio/keymaps/jonavin/keymap.c:257]: just a heads up, snprintf will use u about 1.5kb of firmware space, when it's used.

Comment (Jonavin) [reply keyboards/mechwild/mercutio/keymaps/jonavin/keymap.c:257]: Do you recommend using strncat here instead?

**Review by fauxpark — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!

**Pull request merged**